src/Compilers/Core/AnalyzerDriver/AnalyzerDriverHelper.cs

```diff
@@ -13,76 +13,63 @@ namespace Microsoft.CodeAnalysis.Diagnostics
 {
     internal class AnalyzerDriverHelper
     {
-        internal const string DiagnosticId = "AD0001";
+        private const string DiagnosticId = "AD0001";
         private const string DiagnosticCategory = "Compiler";
```

Comment:
AnalyzerDriverHelper is a type in shared AnalyzerDriver project, which is compiled into both CodeAnalysis.csproj and Features.csproj.
So basically everything that needs to be shared between IDE and compiler driver, such as core Execute*Actions method that just call into the analyzer and handle exceptions are needed by both drivers and lie in AnalyzerDriverHelper.
AnalyzerDriver is on the other hand something tied closely to "compiler" analyzer driver, is tightly coupled with event queue model and a compilation.
We created the shared AnalyzerDriver project to ensure that we don't end up exposing public APIs, just to match current IDE driver's design. Eventually, the goal is to remove this shared project and make everything related to driving analyzer execution move back into AnalyzerDriver, but that needs IDE driver to be event based.
We should surely do a bunch of refactoring suggestions you have given in this review. But I think we should wait for IDE driver to move to event queue before we do those.